### ISSUE TITLE
The CAM methods about gradient failed to run

### ISSUE THREAD

**Issue by liuyx599** [seppia978/ADCC#1]
```python
def ScoreCAM_extracor(image,model, classidx=None):  
    scam=CAMS.GradCAM(model)    # CAM  ScoreCAM, ISCAM, SSCAM is ok    
     
    with torch.no_grad(): out = model(image)
```
when I change the  `scam=CAMS.ScoreCAM(model)` to   `scam=CAMS.GradCAM(model)  `
There is an error occuring:  cannot register a hook on a tensor that doesn't require gradient

The same error occurred in 'GradCAM', 'GradCAMpp', 'SmoothGradCAMpp', 'XGradCAM', 'LayerCAM'
Either CAM  ScoreCAM, ISCAM, SSCAM is ok

**Comment by seppia978** [comment]
It seems that the input tensor you're passing to the extractor isn't requiring the gradient, while for all of those CAM methods gradient is fundamental. I suggest to check it.

**Comment by liuyx599** [comment]
Alas, I'm so stupid. I haven't been able to solve it all day. 
I ensure each tensor's  `requires_grad`  is True

**Comment by liuyx599** [comment]
It's really metaphysics
When I specify a specific convolution layer, it can be used
Thank you very much and wish you a happy life

```python
scam=CAMS.GradCAMpp(model, 'layer4')
```

**Comment by seppia978** [comment]
Yeah, I don't know if you're familiar with CAM based methods but they all work more or less like this: they take the N activations output from a convolutional layer (normally the last one, so the 4th for the resnet-like nets, feature29 fog vgg16 and others) and they linearly combine them with a set of N weights to compute the final saliency map. They way in which they differ is in how they compute those weights.
So they all need a layer to hook to get those activations. Also some of them use the gradient in their algoritm.

But if you'd like to read more just take a look at my publication about this metric, where I explain these concept a little more in detail!

I wish you a good one and a happy life too!
